ironfish/src/memPool/memPool.ts

```diff
@@ -180,47 +231,47 @@ export class MemPool {
     if (deletedTransactions) {
       this.logger.debug(`Deleted ${deletedTransactions} transactions`)
     }
-
     this.head = block.header
+    this.recentlyEvictedCache.flush(this.head.sequence)
   }
 
   async onDisconnectBlock(block: Block): Promise<void> {
-    let addedTransactions = 0
-
     for (const transaction of block.transactions) {
       if (transaction.isMinersFee()) {
         continue
       }
 
-      const added = this.addTransaction(transaction)
-      if (added) {
-        addedTransactions++
-      }
+      this.addTransaction(transaction)
     }
 
-    this.logger.debug(`Added ${addedTransactions} transactions`)
-
     this.head = await this.chain.getHeader(block.header.previousBlockHash)
   }
 
+  /**
+   * Add a new transaction to the mempool. If the mempool is full, the lowest feeRate transactions
+   * are evicted from the mempool.
+   *
+   * Transactions with duplicate nullifers are rejected.
+   *
+   * @param transaction the transaction to add
+   *
+   * @returns true if the transaction is valid to be added. This will STILL return true
+   * even if the transaction doesn't make it into the mempool because of size constraints
+   */
   private addTransaction(transaction: Transaction): boolean {
     const hash = transaction.hash()
 
-    if (this.transactions.has(hash)) {
+    if (this.exists(hash)) {
       return false
     }
 
     // Don't allow transactions with duplicate nullifiers
+    // TODO(daniel): Don't delete transactions if we aren't going to add the transaction anyway
     for (const spend of transaction.spends) {
-      if (this.nullifiers.has(spend.nullifier)) {
-        const existingTransactionHash = this.nullifiers.get(spend.nullifier)
-        Assert.isNotUndefined(existingTransactionHash)
-
-        const existingTransaction = this.transactions.get(existingTransactionHash)
-        if (!existingTransaction) {
-          continue
-        }
+      const existingHash = this.nullifiers.get(spend.nullifier)
+      const existingTransaction = existingHash && this.transactions.get(existingHash)
 
+      if (existingTransaction) {
```

Comment:
I think I did prefer flattening this with a `continue` rather than changing it to an `if`. Also definitely helps to keep refactors out of these PRs

Reply:
yea will change it back